platform/macos/joypad_macos.mm

```diff
@@ -0,0 +1,570 @@
+		// these...
+		_weakify(self);
+		_weakify(controller);
+
+		controller.extendedGamepad.valueChangedHandler = ^(GCExtendedGamepad *gamepad, GCControllerElement *element) {
+			_strongify(self);
+			_strongify(controller);
+
+			int joy_id = [self getJoyIdForController:controller];
+
+			if (element == gamepad.buttonA) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::A,
+						gamepad.buttonA.isPressed);
+			} else if (element == gamepad.buttonB) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::B,
+						gamepad.buttonB.isPressed);
+			} else if (element == gamepad.buttonX) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::X,
+						gamepad.buttonX.isPressed);
+			} else if (element == gamepad.buttonY) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::Y,
+						gamepad.buttonY.isPressed);
+			} else if (element == gamepad.leftShoulder) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::LEFT_SHOULDER,
+						gamepad.leftShoulder.isPressed);
+			} else if (element == gamepad.rightShoulder) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::RIGHT_SHOULDER,
+						gamepad.rightShoulder.isPressed);
+			} else if (element == gamepad.dpad) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_UP,
+						gamepad.dpad.up.isPressed);
+				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_DOWN,
+						gamepad.dpad.down.isPressed);
+				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_LEFT,
+						gamepad.dpad.left.isPressed);
+				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_RIGHT,
+						gamepad.dpad.right.isPressed);
+			} else if (element == gamepad.leftThumbstickButton) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::LEFT_STICK,
+						gamepad.leftThumbstickButton.isPressed);
+			} else if (element == gamepad.rightThumbstickButton) {
+				Input::get_singleton()->joy_button(joy_id, JoyButton::RIGHT_STICK,
+						gamepad.rightThumbstickButton.isPressed);
+			}
+
+			if (element == gamepad.leftThumbstick) {
+				float value = gamepad.leftThumbstick.xAxis.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::LEFT_X, value);
+				value = -gamepad.leftThumbstick.yAxis.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::LEFT_Y, value);
+			} else if (element == gamepad.rightThumbstick) {
+				float value = gamepad.rightThumbstick.xAxis.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::RIGHT_X, value);
+				value = -gamepad.rightThumbstick.yAxis.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::RIGHT_Y, value);
+			} else if (element == gamepad.leftTrigger) {
+				float value = gamepad.leftTrigger.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::TRIGGER_LEFT, value);
+			} else if (element == gamepad.rightTrigger) {
+				float value = gamepad.rightTrigger.value;
+				Input::get_singleton()->joy_axis(joy_id, JoyAxis::TRIGGER_RIGHT, value);
+			}
+
```

Comment:
I have created [macos_xbox-branch](https://github.com/godotengine/godot/compare/master...Alex2782:godot:macos_xbox) on my account

These are all extensions: https://github.com/godotengine/godot/commit/70bb7a3c88269b86a0655dce4fe27f0737fb429d

None: `leftThumbstickButton` and `rightThumbstickButton` require macOS 10.14.1+
The two if-queries above would have to be removed again.